crates/hir_def/src/attr.rs

```diff
@@ -507,6 +559,59 @@ impl AttrSourceMap {
     }
 }
 
+/// A struct to map text ranges from [`Documentation`] back to TextRanges in the syntax tree.
+pub struct DocsRangeMap {
+    source: Vec<InFile<Either<ast::Attr, ast::Comment>>>,
+    // (docstring-line-range, attr_index, attr-string-range)
+    // a mapping from the text range of a line of the [`Documentation`] to the attribute index and
+    // the original (untrimmed) syntax doc line
+    mapping: Vec<(Range<usize>, u32, Range<usize>)>,
+}
+
+impl DocsRangeMap {
+    pub fn map(&self, range: Range<usize>) -> Option<InFile<TextRange>> {
+        let found = self
+            .mapping
+            .binary_search_by(|(probe, ..)| {
```

Comment:
This could use https://docs.rs/text-size/1.1.0/text_size/struct.TextRange.html#method.ordering it seems